### PR TITLE
🎉 New Source: Outreach

### LINKED ISSUE #3627: Outreach.io as  source of data
## Tell us about the new integration you’d like to have
Would love to be able to ingest outreach.io data  ([API doc](https://api.outreach.io/api/v2/docs))into our Snowflake db so we can join outbound campaign data with product usage in our analysis and modeling.

This is something that Stitch supports but not Fivetran.

## Describe the context around this new integration
* Sales teams wants data from outreach about which prospects is getting targeted by the BDR teams and wants to know which campaign they are receiving and finally, when was the last time this contact was contacted.

This will be on a fast sync time. 5-10min.
Would need to support all of the endponts in Outreach API ([docs](https://api.outreach.io/api/v2/docs))

## Describe the alternative you are considering or using
Will have to use Stitch as it's the only ETL provider that support it right now

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200368060087111/1200387048480264) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by lgomezm**
## What
* Adding `Outreach` as new source. Fixes https://github.com/airbytehq/airbyte/issues/3627
* Includes a few streams to start with.

## How
* Implements OAuth authentication.
* Supports both full refresh and incremental syncs.
* Supported streams: `prospects`, `sequences`, `sequence_states`.
* New streams can be added easily.

## Recommended reading order
1. `spec.json`.
2. Schemas in `schemas/*.json`.
3. `source.py`

### Integration tests
Screen Shot 2021-10-31 at 9 20 12 PM

### Unit tests
Screen Shot 2021-10-31 at 9 23 01 PM

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Comment by harshithmullapudi** [comment]
Hey @lgomezm thanks for the contribution 🔥. Can you run `./gradlew format`

File: airbyte-integrations/connectors/source-outreach/source_outreach/source.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    ) -> MutableMapping[str, Any]:
+        params = super().request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
+        if 'updatedAt' in stream_state:
+            params["filter[updatedAt]"] = stream_state['updatedAt'] + '..inf'
+        return params
+
+
+class Prospects(IncrementalOutreachStream):
+    primary_key = "id"
+    cursor_field = "attributes/properties/updatedAt"
```
Comment (harshithmullapudi) [inline airbyte-integrations/connectors/source-outreach/source_outreach/source.py:83]: I think this is not needed as it is already declared in parent class

File: airbyte-integrations/connectors/source-outreach/source_outreach/source.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    def __init__(
+        self,
+        authenticator: HttpAuthenticator,
+        start_date: str = None,
+        **kwargs,
+    ):
+        self.start_date = start_date
+        super().__init__(authenticator=authenticator)
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
```
Comment (harshithmullapudi) [inline airbyte-integrations/connectors/source-outreach/source_outreach/source.py:32]: can you add documentation URL here

File: airbyte-integrations/connectors/source-outreach/source_outreach/source.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+class Prospects(IncrementalOutreachStream):
+    primary_key = "id"
+    cursor_field = "attributes/properties/updatedAt"
+
+    def path(self, **kwargs) -> str:
+        return "prospects"
+
+
+class Sequences(IncrementalOutreachStream):
```
Comment (harshithmullapudi) [inline airbyte-integrations/connectors/source-outreach/source_outreach/source.py:89]: can you add documentation URL here for this stream

File: airbyte-integrations/connectors/source-outreach/source_outreach/source.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    primary_key = "id"
+    cursor_field = "attributes/properties/updatedAt"
+
+    def path(self, **kwargs) -> str:
+        return "sequences"
+
+
+class SequenceStates(IncrementalOutreachStream):
```
Comment (harshithmullapudi) [inline airbyte-integrations/connectors/source-outreach/source_outreach/source.py:97]: can you add documentation URL here for this stream

File: airbyte-integrations/connectors/source-outreach/source_outreach/source.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        payload['redirect_uri'] = self.redirect_uri
+        return payload
+
+
+# Source
+class SourceOutreach(AbstractSource):
+    def _create_authenticator(self, config):
+        return OutreachAuthenticator(
+            redirect_uri=config["redirect_uri"],
+            token_refresh_endpoint="https://api.outreach.io/oauth/token",
```
Comment (harshithmullapudi) [inline airbyte-integrations/connectors/source-outreach/source_outreach/source.py:130]: if this is constant can you declare a variable on the top and use it here

**Review by harshithmullapudi — CHANGES_REQUESTED** [review]
Small changes otherwise LGTM

**Comment by lgomezm** [comment]
Hi @harshithmullapudi. I've just addressed your comments. Please take a look again when you get a chance. Thanks for reviewing!

**Review by harshithmullapudi — APPROVED** [review]

**Comment by harshithmullapudi** [comment]
Hey @lgomezm due to some restriction on Outreach end our credentials are returning 429 in later tests. Do you have some take on this ?

**Comment by lgomezm** [comment]
> Hey @lgomezm due to some restriction on Outreach end our credentials are returning 429 in later tests. Do you have some take on this ?
Hi @harshithmullapudi. Yes, unfortunately it fails intermittently beacuse of `HTTP 429: Too many requests`. One work around I use to check failed scenarios is run separately those cases that failed because of that error, with something like `$ python -m pytest integration_tests -p integration_tests.acceptance -k 'test_two_sequential_reads'`.  That command  runs the test case called `test_two_sequential_reads`.

**Comment by marcosmarxm** [comment]
Hello @lgomezm ! Sorry to not finished your contribution before the date stipulated in the contest. All contributions made before 15-November are eligible to receive the award. We're trying to review your contribution as soon as possible.

**Pull request merged**